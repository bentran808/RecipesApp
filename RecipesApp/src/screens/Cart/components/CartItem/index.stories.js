import { storiesOf } from '@storybook/react-native';
import { recipe } from 'mocks';
import React from 'react';
import CenterView from 'storybook/stories/CenterView';
import CartItem from '.';

storiesOf('Cart Item', module)
  .addDecorator((getStory) => <CenterView>{getStory()}</CenterView>)
  .add('default', () => (
    <CartItem
      item={{
        item: recipe,
        quantity: 1,
        inCart: true
      }}
    />
  ));
