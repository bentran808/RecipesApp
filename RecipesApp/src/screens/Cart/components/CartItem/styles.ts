import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    backgroundColor: 'white',
    margin: 10,
    flex: 1
  },
  photoContainer: { padding: 5 },
  photo: {
    width: 75,
    height: 75,
    borderRadius: 5
  },
  infoContainer: {
    flex: 1,
    padding: 10
  },
  infoTitle: { fontSize: 18 },
  priceContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  priceText: {
    fontSize: 16,
    fontWeight: 'bold'
  },
  quantityContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    alignSelf: 'flex-end',
    borderWidth: 1,
    borderColor: 'red',
    borderRadius: 5
  },
  quantityText: { color: 'red' },
  actionText: {
    color: 'white',
    paddingHorizontal: 10,
    fontWeight: '600'
  },
  rightAction: {
    backgroundColor: 'red',
    justifyContent: 'center'
  }
});

export default styles;
