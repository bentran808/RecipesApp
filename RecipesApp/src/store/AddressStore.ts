import { toJS } from 'mobx';
import { types, destroy, SnapshotOut, getRoot, applySnapshot } from 'mobx-state-tree';
import { RootStore } from 'store/store';

const AddressEntry = types
  .model('AddressEntry', {
    id: types.optional(types.identifierNumber, 0),
    type: types.optional(types.string, ''),
    address: types.optional(types.string, '')
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).address.removeItem(self);
    },
    edit(type: string, address: string) {
      self.type = type;
      self.address = address;
    }
  }));

export type AddressModel = SnapshotOut<typeof AddressEntry>;

const AddressStore = types
  .model('', {
    items: types.array(AddressEntry),
    itemUsing: types.optional(AddressEntry, {})
  })
  .volatile(() => ({
    address: ''
  }))
  .actions((self) => ({
    setAddress(address: string) {
      self.address = address;
    },
    addNewItem(type: string, address: string) {
      const id = self.items.reduce((maxId, item) => Math.max(item.id, maxId), -1) + 1;
      self.items.push({ id, type, address });
    },
    removeItem(item: AddressModel) {
      destroy(item);
    },
    existAddress(address?: AddressModel) {
      return self.items.find((item) => item.id === address?.id);
    },
    setItemUsing(item: AddressModel) {
      applySnapshot(self, {
        ...self,
        itemUsing: item
      });
    }
  }))
  .views((self) => ({
    get addresses() {
      return toJS(self.items);
    }
  }));

export default AddressStore;
