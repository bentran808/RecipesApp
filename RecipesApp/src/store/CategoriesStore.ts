import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';
import { transformCategories } from 'utils';
import { RecipesEntry } from './RecipesStore';

export const CategoriesEntry = types.model('CategoriesEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string,
  recipes: types.maybe(types.array(RecipesEntry))
});

export type CategoryModel = SnapshotOut<typeof CategoriesEntry>;

const CategoriesStore = types
  .model('CategoriesStore', {
    lists: types.optional(types.array(CategoriesEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get categories() {
      return self.lists;
    },
    get categoriesJS() {
      return transformCategories(toJS(this.categories));
    }
  }))
  .actions((self) => ({
    fetchCategories: flow(function* fetchCategories() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchCategoriesRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        console.log(error);
        self.state = 'error';
      }
    })
  }));

export default CategoriesStore;
