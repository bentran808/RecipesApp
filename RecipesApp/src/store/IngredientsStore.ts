import { recipesApi } from 'api';
import { toJS } from 'mobx';
import { flow, SnapshotOut, types } from 'mobx-state-tree';

export const IngredientsEntry = types.model('IngredientsEntry', {
  id: types.number,
  name: types.string,
  photo_url: types.string
});

export type IngredientModel = SnapshotOut<typeof IngredientsEntry>;

const IngredientsStore = types
  .model('IngredientsStore', {
    lists: types.optional(types.array(IngredientsEntry), []),
    state: types.optional(types.enumeration('State', ['pending', 'done', 'error']), 'done')
  })
  .views((self) => ({
    get ingredients() {
      return self.lists;
    },
    get ingredientsJS() {
      return toJS(this.ingredients);
    }
  }))
  .actions((self) => ({
    fetchIngredients: flow(function* fetchIngredients() {
      self.state = 'pending';
      try {
        const response = yield recipesApi.fetchIngredientsRequest();
        self.lists = response.data;
        self.state = 'done';
      } catch (error) {
        self.state = 'error';
      }
    })
  }));

export default IngredientsStore;
