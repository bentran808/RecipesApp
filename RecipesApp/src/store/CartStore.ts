import { toJS } from 'mobx';
import { applySnapshot, destroy, getRoot, SnapshotOut, types } from 'mobx-state-tree';
import { RecipeModel, RecipesEntry } from 'store/RecipesStore';
import { RootStore } from 'store/store';

export const CartEntry = types
  .model('CartEntry', {
    item: RecipesEntry,
    quantity: types.optional(types.number, 1),
    inCart: types.optional(types.boolean, true)
  })
  .actions((self) => ({
    remove() {
      getRoot<typeof RootStore>(self).cart.removeItem(self);
    },
    increase(quantity?: number) {
      self.quantity = quantity ? self.quantity + quantity : ++self.quantity;
    },
    decrease() {
      self.quantity === 1 ? this.remove() : --self.quantity;
    },
    toggle() {
      self.inCart = !self.inCart;
    }
  }));

const BillEntry = types.model('BillEntry', {
  type: types.optional(types.string, ''),
  name: types.optional(types.string, ''),
  code: types.optional(types.string, ''),
  price: types.optional(types.number, 0)
});

const OrderedEntry = types.model('OrderedEntry', {
  items: types.array(
    types.model({
      name: types.optional(types.string, ''),
      quantity: types.optional(types.number, 0)
    })
  ),
  total: types.number,
  createdAt: types.string
});

export type CartModel = SnapshotOut<typeof CartEntry>;
export type BillModel = SnapshotOut<typeof BillEntry>;
export type OrderedModel = SnapshotOut<typeof OrderedEntry>;

const CartStore = types
  .model('CartStore', {
    items: types.array(CartEntry),
    billItems: types.array(BillEntry),
    orderedItems: types.array(OrderedEntry)
  })
  .views((self) => ({
    get recipes() {
      return toJS(self.items);
    },
    get recipesInCart() {
      return this.recipes.filter((item) => item.inCart);
    },
    get inCartCount() {
      return this.recipesInCart.length;
    },
    get total() {
      return this.recipesInCart.reduce((sum, item) => sum + item.quantity * 20, 0);
    },
    get billDetails() {
      return toJS(self.billItems);
    },
    get toPay() {
      return this.billDetails.reduce(
        (sum, item) => (item.type === 'discount' ? sum - item.price : sum + item.price),
        0
      );
    },
    get orderedList() {
      return toJS(self.orderedItems);
    }
  }))
  .volatile(() => ({
    discountInput: '',
    invalidCoupon: false
  }))
  .actions((self) => ({
    addToCart(item: RecipeModel, quantity?: number) {
      const itemInCart = this.hasInCart(item);
      itemInCart ? itemInCart.increase(quantity) : self.items.push({ item, quantity });
    },
    removeItem(item: CartModel) {
      destroy(item);
    },
    hasInCart(recipe: RecipeModel) {
      return self.items.find((item) => item.item.id === recipe.id);
    },
    setDiscountInput(text: string) {
      if (!text) {
        self.invalidCoupon = false;
      }
      self.discountInput = text;
    },
    applyCoupon(item: BillModel) {
      const existItem = self.billItems.findIndex((bill) => bill.type === 'discount');
      if (existItem > -1) {
        self.billItems.splice(existItem, 1, item);
      } else {
        self.billItems.push(item);
      }
    },
    payment(order: OrderedModel) {
      self.orderedItems.unshift(order);
      applySnapshot(self, {
        ...self,
        items: [],
        billItems: []
      });
    },
    isInvalidCoupon(status: boolean) {
      self.invalidCoupon = status;
    }
  }));

export default CartStore;
