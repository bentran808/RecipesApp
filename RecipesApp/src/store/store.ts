import { Instance, types } from 'mobx-state-tree';
import RecipesStore from 'store/RecipesStore';
import CategoriesStore from './CategoriesStore';

const RootStore = types.model('RootStore', {
  categories: CategoriesStore,
  recipes: RecipesStore
});

const store = RootStore.create({
    categories: {},
    recipes: {}
});

export type RootStoreModel = Instance<typeof RootStore>;

export default store;
