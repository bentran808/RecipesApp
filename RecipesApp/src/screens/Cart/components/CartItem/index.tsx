import Button from 'components/Button';
import { useStore } from 'context';
import { observer } from 'mobx-react-lite';
import React, { useCallback } from 'react';
import isEqual from 'react-fast-compare';
import { Alert, Animated, Image, Text, View } from 'react-native';
import { RectButton } from 'react-native-gesture-handler';
import Swipeable from 'react-native-gesture-handler/Swipeable';
import { CartModel } from 'store/CartStore';
import styles from './styles';

type Props = {
  item: CartModel;
};

const CartItem = ({ item }: Props) => {
  const { cart } = useStore();

  const handleIncrease = useCallback(() => {
    const itemInCart = cart.hasInCart(item.item);
    itemInCart?.increase && itemInCart.increase();
  }, [item]);

  const handleDecrease = useCallback(() => {
    const itemInCart = cart.hasInCart(item.item);
    itemInCart?.decrease && itemInCart.decrease();
  }, [item]);

  const handleConfirmDelete = useCallback(() => {
    const itemInCart = cart.hasInCart(item.item);
    itemInCart?.remove && itemInCart.remove();
  }, [item]);

  const handlePressDelete = () => {
    Alert.alert('Confirm', 'Are you sure?', [
      {
        text: 'Cancel'
      },
      { text: 'OK', onPress: handleConfirmDelete }
    ]);
  };

  const renderRightActions = (progress: any, dragX: any) => {
    const scale = dragX.interpolate({
      inputRange: [-40, 0],
      outputRange: [0.7, 0]
    });
    return (
      <RectButton style={styles.rightAction} onPress={handlePressDelete}>
        <Animated.Text
          style={{
            ...styles.actionText,
            transform: [{ scale }]
          }}
        >
          Delete
        </Animated.Text>
      </RectButton>
    );
  };

  return (
    <Swipeable renderRightActions={renderRightActions} containerStyle={{ width: '100%' }}>
      <View style={styles.container}>
        <View style={styles.photoContainer}>
          <Image
            style={styles.photo}
            source={{
              uri: item.item.photo_url
            }}
          />
        </View>
        <View style={styles.infoContainer}>
          <Text style={styles.infoTitle}>{item.item.title}</Text>
          <View style={styles.priceContainer}>
            <Text style={styles.priceText}>$20</Text>
            <View style={styles.quantityContainer}>
              <Button
                testID="decreaseBtn"
                color="red"
                title="-"
                onPress={handleDecrease}
                fontSize={20}
              />
              <Text style={styles.quantityText}>{item.quantity}</Text>
              <Button
                testID="increaseBtn"
                color="red"
                title="+"
                onPress={handleIncrease}
                fontSize={20}
              />
            </View>
          </View>
        </View>
      </View>
    </Swipeable>
  );
};

export default React.memo(observer(CartItem), isEqual);
