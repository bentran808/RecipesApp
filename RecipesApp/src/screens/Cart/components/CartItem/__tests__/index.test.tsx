import { fireEvent, render } from '@testing-library/react-native';
import { StoreProvider } from 'context';
import { recipe } from 'mocks';
import React from 'react';
import { Alert } from 'react-native';
import renderer from 'react-test-renderer';
import store from 'store/store';
import CartItem from '..';

const spyAlert: any = jest.spyOn(Alert, 'alert');

describe('Cart Item Component', () => {
  const props = {
    item: {
      item: recipe,
      quantity: 1,
      inCart: true
    }
  };

  test('should render correctly', () => {
    const tree = renderer
      .create(
        <StoreProvider value={store}>
          <CartItem {...props} />
        </StoreProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should call function handleIncrease', () => {
    store.cart.hasInCart = jest.fn();
    const { getByTestId } = render(
      <StoreProvider value={store}>
        <CartItem {...props} />
      </StoreProvider>
    );
    const button = getByTestId('increaseBtn');

    fireEvent.press(button);

    expect(store.cart.hasInCart).toHaveBeenCalled();
  });

  test('should call function handleDecrease', () => {
    store.cart.hasInCart = jest.fn();
    const { getByTestId } = render(
      <StoreProvider value={store}>
        <CartItem {...props} />
      </StoreProvider>
    );
    const button = getByTestId('decreaseBtn');

    fireEvent.press(button);

    expect(store.cart.hasInCart).toHaveBeenCalled();
  });

  test('should call function handlePressDelete', () => {
    store.cart.hasInCart = jest.fn();
    const { getByText } = render(
      <StoreProvider value={store}>
        <CartItem {...props} />
      </StoreProvider>
    );

    const button = getByText('Delete');
    fireEvent.press(button);
    expect(Alert.alert).toHaveBeenCalled();

    spyAlert.mock.calls[0][2][1].onPress();
    expect(store.cart.hasInCart).toHaveBeenCalled();
  });
});
