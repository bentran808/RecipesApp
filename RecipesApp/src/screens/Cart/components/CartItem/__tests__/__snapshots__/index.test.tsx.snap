// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Item Component should render correctly 1`] = `
<View
  collapsable={false}
  forwardedRef={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  style={
    Object {
      "overflow": "hidden",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "flexDirection": "row-reverse",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateX": -10000,
          },
        ],
      }
    }
  >
    <RNGestureHandlerButton
      activeOpacity={0.105}
      collapsable={false}
      onActiveStateChange={[Function]}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "red",
          "justifyContent": "center",
        }
      }
      underlayColor="black"
    >
      <View
        style={
          Object {
            "backgroundColor": "black",
            "borderBottomLeftRadius": undefined,
            "borderBottomRightRadius": undefined,
            "borderRadius": undefined,
            "borderTopLeftRadius": undefined,
            "borderTopRightRadius": undefined,
            "bottom": 0,
            "left": 0,
            "opacity": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "600",
            "paddingHorizontal": 10,
            "transform": Array [
              Object {
                "scale": 0,
              },
            ],
          }
        }
      >
        Delete
      </Text>
    </RNGestureHandlerButton>
    <View
      onLayout={[Function]}
    />
  </View>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    pointerEvents="auto"
    style={
      Object {
        "transform": Array [
          Object {
            "translateX": -0,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "flexDirection": "row",
          "margin": 10,
        }
      }
    >
      <View
        style={
          Object {
            "padding": 5,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://www.texanerin.com/content/uploads/2019/06/nobake-chocolate-cookies-1-650x975.jpg",
            }
          }
          style={
            Object {
              "borderRadius": 5,
              "height": 75,
              "width": 75,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 18,
            }
          }
        >
          Oatmeal Cookies
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            $20
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "flex-end",
                "borderColor": "red",
                "borderRadius": 5,
                "borderWidth": 1,
                "flexDirection": "row",
              }
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="decreaseBtn"
            >
              <View
                style={
                  Array [
                    Object {
                      "borderRadius": 100,
                      "marginHorizontal": 10,
                    },
                    Object {},
                    undefined,
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "red",
                        "fontSize": 20,
                        "fontWeight": "normal",
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  -
                </Text>
              </View>
            </View>
            <Text
              style={
                Object {
                  "color": "red",
                }
              }
            >
              1
            </Text>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID="increaseBtn"
            >
              <View
                style={
                  Array [
                    Object {
                      "borderRadius": 100,
                      "marginHorizontal": 10,
                    },
                    Object {},
                    undefined,
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "red",
                        "fontSize": 20,
                        "fontWeight": "normal",
                        "textAlign": "center",
                      },
                    ]
                  }
                >
                  +
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
